File: src/desktop/ExcelDesktop/Services/ChartingEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Drawing;
using Microsoft.ExcelDesktop.Interfaces;
using Microsoft.ExcelDesktop.Models;
using Microsoft.ExcelDesktop.Utils;

namespace Microsoft.ExcelDesktop.Services
{
    /// <summary>
    /// Implements the IChartingEngine interface to provide charting functionality for Excel Desktop
    /// </summary>
    public class ChartingEngine : IChartingEngine
    {
        private readonly Dictionary<ChartType, Func<Chart>> chartFactories;

        /// <summary>
        /// Initializes a new instance of the ChartingEngine class
        /// </summary>
        public ChartingEngine()
        {
            // Initialize the chartFactories dictionary with supported chart types and their respective factory methods
            chartFactories = new Dictionary<ChartType, Func<Chart>>
            {
                { ChartType.Bar, () => new Chart { Type = ChartType.Bar } },
                { ChartType.Line, () => new Chart { Type = ChartType.Line } },
                { ChartType.Pie, () => new Chart { Type = ChartType.Pie } },
                // Add more chart types as needed
            };
        }

        /// <summary>
        /// Creates a new chart based on the specified data range and chart type
        /// </summary>
        /// <param name="worksheet">The worksheet containing the data</param>
        /// <param name="dataRange">The range of cells containing the data for the chart</param>
        /// <param name="chartType">The type of chart to create</param>
        /// <returns>The newly created chart object</returns>
        public Chart CreateChart(Worksheet worksheet, string dataRange, ChartType chartType)
        {
            // Validate the input parameters
            if (worksheet == null)
                throw new ArgumentNullException(nameof(worksheet));
            if (string.IsNullOrWhiteSpace(dataRange))
                throw new ArgumentException("Data range cannot be empty", nameof(dataRange));
            if (!chartFactories.ContainsKey(chartType))
                throw new ArgumentException("Unsupported chart type", nameof(chartType));

            // Extract data from the specified range using CellAddressHelper
            var data = CellAddressHelper.ExtractDataFromRange(worksheet, dataRange);

            // Validate the extracted data using DataValidator
            if (!DataValidator.ValidateChartData(data))
                throw new InvalidOperationException("Invalid data for chart creation");

            // Create a new chart of the specified type using the chart factory
            var chart = chartFactories[chartType]();

            // Populate the chart with the extracted data
            PopulateChartWithData(chart, data);

            // Apply default formatting to the chart
            ApplyDefaultFormatting(chart);

            return chart;
        }

        /// <summary>
        /// Updates an existing chart with new data or settings
        /// </summary>
        /// <param name="chart">The chart to update</param>
        /// <param name="worksheet">The worksheet containing the new data</param>
        /// <param name="newDataRange">The new range of cells containing the data for the chart</param>
        public void UpdateChart(Chart chart, Worksheet worksheet, string newDataRange)
        {
            // Validate the input parameters
            if (chart == null)
                throw new ArgumentNullException(nameof(chart));
            if (worksheet == null)
                throw new ArgumentNullException(nameof(worksheet));
            if (string.IsNullOrWhiteSpace(newDataRange))
                throw new ArgumentException("New data range cannot be empty", nameof(newDataRange));

            // Extract data from the new range using CellAddressHelper
            var newData = CellAddressHelper.ExtractDataFromRange(worksheet, newDataRange);

            // Validate the extracted data using DataValidator
            if (!DataValidator.ValidateChartData(newData))
                throw new InvalidOperationException("Invalid data for chart update");

            // Update the chart's data series with the new data
            UpdateChartData(chart, newData);

            // Refresh the chart's layout and appearance
            RefreshChartLayout(chart);
        }

        /// <summary>
        /// Changes the type of an existing chart
        /// </summary>
        /// <param name="chart">The chart to modify</param>
        /// <param name="newChartType">The new chart type</param>
        public void ChangeChartType(Chart chart, ChartType newChartType)
        {
            // Validate if the new chart type is compatible with the current data
            if (!IsChartTypeCompatible(chart, newChartType))
                throw new InvalidOperationException("The new chart type is not compatible with the current data");

            // Create a new chart of the specified type using the chart factory
            var newChart = chartFactories[newChartType]();

            // Transfer data from the old chart to the new chart
            TransferChartData(chart, newChart);

            // Copy relevant formatting and customization settings
            CopyChartSettings(chart, newChart);

            // Replace the old chart with the new chart in the worksheet
            ReplaceChartInWorksheet(chart, newChart);
        }

        /// <summary>
        /// Applies formatting options to a chart
        /// </summary>
        /// <param name="chart">The chart to format</param>
        /// <param name="formatOptions">The formatting options to apply</param>
        public void FormatChart(Chart chart, ChartFormatOptions formatOptions)
        {
            // Validate the input parameters
            if (chart == null)
                throw new ArgumentNullException(nameof(chart));
            if (formatOptions == null)
                throw new ArgumentNullException(nameof(formatOptions));

            // Apply the specified formatting options to the chart
            ApplyChartFormatting(chart, formatOptions);

            // Update chart title, axis labels, and legend as needed
            UpdateChartLabels(chart, formatOptions);

            // Apply color scheme and other visual properties
            ApplyChartVisualProperties(chart, formatOptions);

            // Refresh the chart's appearance
            RefreshChartAppearance(chart);
        }

        /// <summary>
        /// Exports a chart to a specified file format
        /// </summary>
        /// <param name="chart">The chart to export</param>
        /// <param name="filePath">The file path where the chart will be saved</param>
        /// <param name="format">The export format</param>
        /// <returns>True if the export was successful, false otherwise</returns>
        public bool ExportChart(Chart chart, string filePath, ExportFormat format)
        {
            // Validate the file path and format
            if (string.IsNullOrWhiteSpace(filePath))
                throw new ArgumentException("File path cannot be empty", nameof(filePath));

            try
            {
                // Convert the chart to the specified format
                byte[] chartData = ConvertChartToFormat(chart, format);

                // Save the chart to the specified file path
                System.IO.File.WriteAllBytes(filePath, chartData);

                return true;
            }
            catch (Exception ex)
            {
                // Handle any exceptions during the export process
                Console.WriteLine($"Error exporting chart: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Validates the data range for chart creation or update
        /// </summary>
        /// <param name="worksheet">The worksheet containing the data</param>
        /// <param name="dataRange">The range of cells to validate</param>
        /// <returns>True if the data range is valid, false otherwise</returns>
        public bool ValidateDataRange(Worksheet worksheet, string dataRange)
        {
            // Parse the data range string
            var (startCell, endCell) = CellAddressHelper.ParseRange(dataRange);

            // Check if the range is within the worksheet bounds
            if (!CellAddressHelper.IsRangeWithinWorksheet(worksheet, startCell, endCell))
                return false;

            // Verify that the range contains valid data for charting
            var data = CellAddressHelper.ExtractDataFromRange(worksheet, dataRange);
            return DataValidator.ValidateChartData(data);
        }

        /// <summary>
        /// Extracts data from the worksheet for chart creation or update
        /// </summary>
        /// <param name="worksheet">The worksheet containing the data</param>
        /// <param name="dataRange">The range of cells to extract data from</param>
        /// <returns>Extracted data for the chart</returns>
        public ChartData ExtractChartData(Worksheet worksheet, string dataRange)
        {
            // Parse the data range string
            var data = CellAddressHelper.ExtractDataFromRange(worksheet, dataRange);

            // Determine data series and categories
            var chartData = new ChartData
            {
                Categories = data.First().Skip(1).ToList(),
                Series = data.Skip(1).Select(row => new Series
                {
                    Name = row.First(),
                    Values = row.Skip(1).Select(cell => Convert.ToDouble(cell)).ToList()
                }).ToList()
            };

            return chartData;
        }

        // Private helper methods

        private void PopulateChartWithData(Chart chart, List<List<object>> data)
        {
            // Implementation for populating the chart with data
        }

        private void ApplyDefaultFormatting(Chart chart)
        {
            // Implementation for applying default formatting to the chart
        }

        private void UpdateChartData(Chart chart, List<List<object>> newData)
        {
            // Implementation for updating the chart's data
        }

        private void RefreshChartLayout(Chart chart)
        {
            // Implementation for refreshing the chart's layout
        }

        private bool IsChartTypeCompatible(Chart chart, ChartType newChartType)
        {
            // Implementation for checking chart type compatibility
            return true; // Placeholder
        }

        private void TransferChartData(Chart oldChart, Chart newChart)
        {
            // Implementation for transferring data between charts
        }

        private void CopyChartSettings(Chart sourceChart, Chart targetChart)
        {
            // Implementation for copying chart settings
        }

        private void ReplaceChartInWorksheet(Chart oldChart, Chart newChart)
        {
            // Implementation for replacing a chart in the worksheet
        }

        private void ApplyChartFormatting(Chart chart, ChartFormatOptions formatOptions)
        {
            // Implementation for applying formatting options to the chart
        }

        private void UpdateChartLabels(Chart chart, ChartFormatOptions formatOptions)
        {
            // Implementation for updating chart labels
        }

        private void ApplyChartVisualProperties(Chart chart, ChartFormatOptions formatOptions)
        {
            // Implementation for applying visual properties to the chart
        }

        private void RefreshChartAppearance(Chart chart)
        {
            // Implementation for refreshing the chart's appearance
        }

        private byte[] ConvertChartToFormat(Chart chart, ExportFormat format)
        {
            // Implementation for converting the chart to the specified format
            return new byte[0]; // Placeholder
        }
    }

    // Additional types used in the ChartingEngine class
    public enum ChartType
    {
        Bar,
        Line,
        Pie
        // Add more chart types as needed
    }

    public enum ExportFormat
    {
        PNG,
        SVG,
        PDF
        // Add more export formats as needed
    }

    public class ChartFormatOptions
    {
        // Properties for chart formatting options
    }

    public class ChartData
    {
        public List<string> Categories { get; set; }
        public List<Series> Series { get; set; }
    }

    public class Series
    {
        public string Name { get; set; }
        public List<double> Values { get; set; }
    }
}
```

This implementation of the ChartingEngine class provides the core functionality for creating, updating, and managing charts in the Excel Desktop application. It includes methods for creating charts, updating existing charts, changing chart types, formatting charts, and exporting charts to different file formats.

The class uses dependency injection to work with other components like CellAddressHelper and DataValidator. It also implements error handling and input validation to ensure robust operation.

Note that some of the private helper methods have placeholder implementations, as their specific logic would depend on the exact requirements and the implementation of other components in the system.

Human tasks that need to be addressed:

```csharp
// TODO: Implement specific chart creation logic for each supported chart type
// TODO: Develop robust data validation and error handling mechanisms
// TODO: Implement export functionality for various file formats (PNG, SVG, PDF)
// TODO: Optimize chart rendering performance for large datasets
// TODO: Ensure proper memory management and resource disposal
// TODO: Implement unit tests for all public methods of the ChartingEngine class

File: src/desktop/ExcelCore/ChartingEngine.cpp
#include "ChartingEngine.h"
#include "DataStructures.h"
#include <algorithm>
#include <stdexcept>
#include <memory>
#include <vector>

// Constructor implementation for the ChartingEngine class
ChartingEngine::ChartingEngine() {
    // Initialize any necessary resources for chart creation and management
    // Set up any required default configurations
}

// Implementation of the createChart function
std::unique_ptr<Chart> ChartingEngine::createChart(const Worksheet& worksheet, const CellAddress& startCell, const CellAddress& endCell, ChartType type) {
    // Validate input parameters
    if (!isValidRange(startCell, endCell)) {
        throw std::invalid_argument("Invalid cell range for chart creation");
    }

    // Extract data from the specified range in the worksheet
    std::vector<std::vector<double>> chartData = extractDataFromRange(worksheet, startCell, endCell);

    // Create a new Chart object of the specified type using a factory method
    std::unique_ptr<Chart> newChart = createChartByType(type, chartData);

    // Configure the chart with the extracted data
    newChart->setData(chartData);

    // Add the chart to the charts vector
    charts.push_back(newChart.get());

    // Return a unique pointer to the created chart
    return newChart;
}

// Implementation of the updateChart function
bool ChartingEngine::updateChart(Chart& chart, const Worksheet& worksheet, const CellAddress& startCell, const CellAddress& endCell) {
    // Validate input parameters
    if (!isValidRange(startCell, endCell)) {
        return false;
    }

    // Extract new data from the specified range in the worksheet
    std::vector<std::vector<double>> newData = extractDataFromRange(worksheet, startCell, endCell);

    // Update the chart with the new data using chart.setData()
    chart.setData(newData);

    // Recalculate and redraw the chart
    chart.recalculate();
    chart.redraw();

    return true;
}

// Implementation of the deleteChart function
bool ChartingEngine::deleteChart(const Chart& chart) {
    // Find the specified chart in the charts vector using std::find_if
    auto it = std::find_if(charts.begin(), charts.end(),
        [&chart](const Chart* c) { return c == &chart; });

    // If found, use std::vector::erase to remove the chart from the vector
    if (it != charts.end()) {
        charts.erase(it);
        return true;
    }

    return false;
}

// Implementation of the renderChart function
std::vector<uint8_t> ChartingEngine::renderChart(const Chart& chart, RenderFormat format) {
    // Validate the chart and format parameters
    if (!isValidChart(chart) || !isValidRenderFormat(format)) {
        throw std::invalid_argument("Invalid chart or render format");
    }

    // Prepare the rendering context based on the specified format
    RenderContext context(format);

    // Call chart.render() to get the raw chart data
    std::vector<uint8_t> rawData = chart.render(context);

    // Convert the raw data to the specified format if necessary
    std::vector<uint8_t> formattedData = convertToFormat(rawData, format);

    // Return the rendered data as a vector of bytes
    return formattedData;
}

// Helper function to check if a cell range is valid
bool ChartingEngine::isValidRange(const CellAddress& startCell, const CellAddress& endCell) {
    // Implement logic to check if the range is valid
    // For example, check if startCell is before endCell in both row and column
    return true; // Placeholder
}

// Helper function to extract data from a worksheet range
std::vector<std::vector<double>> ChartingEngine::extractDataFromRange(const Worksheet& worksheet, const CellAddress& startCell, const CellAddress& endCell) {
    // Implement logic to extract data from the worksheet
    // This is a placeholder implementation
    return std::vector<std::vector<double>>();
}

// Helper function to create a chart based on its type
std::unique_ptr<Chart> ChartingEngine::createChartByType(ChartType type, const std::vector<std::vector<double>>& data) {
    // Implement a factory method to create different types of charts
    // This is a placeholder implementation
    return std::make_unique<Chart>();
}

// Helper function to check if a chart is valid
bool ChartingEngine::isValidChart(const Chart& chart) {
    // Implement logic to validate the chart
    return true; // Placeholder
}

// Helper function to check if a render format is valid
bool ChartingEngine::isValidRenderFormat(RenderFormat format) {
    // Implement logic to validate the render format
    return true; // Placeholder
}

// Helper function to convert raw data to a specific format
std::vector<uint8_t> ChartingEngine::convertToFormat(const std::vector<uint8_t>& rawData, RenderFormat format) {
    // Implement logic to convert raw data to the specified format
    // This is a placeholder implementation
    return rawData;
}

// Destructor implementation for the ChartingEngine class
ChartingEngine::~ChartingEngine() {
    // Implement memory management and resource cleanup
    for (auto chart : charts) {
        delete chart;
    }
    charts.clear();
}

// Human tasks:
// 1. Implement specific chart creation logic for different chart types
// 2. Add error handling and logging for chart operations
// 3. Implement memory management and resource cleanup in the ChartingEngine destructor
// 4. Optimize data extraction and chart rendering for large datasets
// 5. Implement thread-safety for chart operations if required